community/bolt/src/main/java/org/neo4j/bolt/security/auth/BasicAuthentication.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.bolt.security.auth;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.neo4j.kernel.api.exceptions.Status;
+import org.neo4j.server.security.auth.AuthManager;
+
+/**
+ * Performs basic authentication with user name and password.
+ */
+public class BasicAuthentication implements Authentication
+{
+    private final AuthManager authManager;
+    private final static String SCHEME = "basic";
+
+
+    public BasicAuthentication( AuthManager authManager )
+    {
+        this.authManager = authManager;
+    }
+
+    @Override
+    public void authenticate( Map<String,Object> authToken ) throws AuthenticationException
+    {
+        if ( !SCHEME.equals( authToken.get( SCHEME_KEY ) ) )
+        {
+            throw new AuthenticationException( Status.Security.AuthenticationFailed,
+                    "Authorization token must contain: '" + SCHEME_KEY + " : " + SCHEME + "'" );
+        }
+
+        String user = safeCast( PRINCIPAL, authToken );
+        String password = safeCast( CREDENTIALS, authToken );
+        if ( authToken.containsKey( NEW_CREDENTIALS ) )
+        {
+            update( user, password, safeCast( NEW_CREDENTIALS, authToken ) );
+        }
+        else
+        {
+            authenticate( user, password );
+        }
+    }
+
+    private void authenticate( String user, String password ) throws AuthenticationException
+    {
+        switch ( authManager.authenticate( user, password ) )
+        {
+        case SUCCESS:
+            break;
+        case PASSWORD_CHANGE_REQUIRED:
+            throw new AuthenticationException( Status.Security.CredentialsExpired );
+        case TOO_MANY_ATTEMPTS:
+            throw new AuthenticationException( Status.Security.AuthenticationRateLimit );
+        case FAILURE:
+            throw new AuthenticationException( Status.Security.AuthenticationFailed );
```

Comment:
Shall we add a default catch so that if we added more items in the enums and forgot to update this switch case, we could still catch it with an error?